### ISSUE TITLE
Sculpt OS 24.04

### ISSUE THREAD

**Issue by nfeske** [genodelabs/genode#5174]
This issue is intended for changes on account of the forthcoming Sculpt OS version 24.04.

**Comment by nfeske** [comment]
Commit https://github.com/genodelabs/genode/commit/797d2d47d1fcd30906c2df796b11b9b3f06912c6 bumps the version to 24.04. This means that there will be no ABI changes until the release.

**Comment by nfeske** [comment]
Commit https://github.com/genodelabs/genode/commit/d17aa7d7219181836e4724f7c1482356d25351cd makes the driver selection a little bit more flexible. The PS/2 and Intel drivers can now explicitly be suppressed by using the sculpt-manager config attributes `ps2="no"` and `intel_gpu="no"`. The latter is useful to let the system fall back to VESA or boot-fb.

**Comment by alex-ab** [comment]
Currently boot_fb is not working:
```
[init -> runtime]   provides service Block
[init -> runtime] child "runtime_view" requests resources: cap_quota=9
[init -> runtime] Warning: boot_fb: no route to service "ROM" (label="boot_fb -> platform_info")
[init -> runtime -> boot_fb] Error: ROM-session creation failed (label="platform_info", ram_quota=6K, cap_quota=3, )
[init -> runtime -> boot_fb] Error: could not open ROM session for "platform_info"
[init -> runtime -> boot_fb] Error: Uncaught exception of type 'Genode::Rom_connection::Rom_connection_failed'
[init -> runtime -> boot_fb] Warning: abort called - thread: ep
[init -> runtime] child "usb" announces service "Usb"
``

**Comment by nfeske** [comment]
@alex-ab sorry about the boot_fb issue and thanks for the fix. I presumably tested the boot-fb variant on Monday after my latest driver-policy tweaks. But when using image/uefi with the sculpt_test.run script, Qemu happened to boot the (older) Sculpt OS version present on the ahci image, not the current disk image. The test worked well. ;-) I merged your fix to staging just now.

**Comment by alex-ab** [comment]
@nfeske: all fine, that is the testing phase for ;)

I published close to all packages of mine. Unfortunately, the download of packages into the ram_fs, when used as depot source, reliable stucks with large packets, so I can't rely test the proper working of the packages by now.

**Comment by chelmuth** [comment]
> Unfortunately, the download of packages into the ram_fs, when used as depot source, reliable stucks with large packets, so I can't really test the proper working of the packages by now.

@alex-ab Is this limited to the ram_fs or does it also get stuck with rump?

**Comment by alex-ab** [comment]
> @alex-ab Is this limited to the ram_fs or does it also get stuck with rump?

With rump_fs the download finishes. After several "request resources" of ram_fs it stucks finally.

**Comment by cproc** [comment]
I got the following error on the current staging branch:

```
[init -> runtime -> boot_fb] using boot framebuffer: 800x600x32 @ 0x4000000000 type=1 pitch=3200
Error: I/O memory [0000004000000000,0000004010000000) not available
Error: Local MMIO mapping failed!
[init -> drivers -> platform_drv] Error: IO_MEM-session creation failed (label="", ram_quota=6K, cap_quota=3, base=0x4000000000, size=0x10000000, wc=yes)
[init -> drivers -> platform_drv] Error: Uncaught exception of type 'Genode::Service_denied'
[init -> drivers -> platform_drv] Warning: abort called - thread: ep
```

**Comment by chelmuth** [comment]
> 1000

The question is: Why does platform_drv also request the framebuffer IO memory from core? boot_fb does not use the platform_drv so it must be another component.

**Comment by jschlatow** [comment]
@cproc On what hardware have you seen this? Could you please post or email me the `/report/drivers/devices` (running the last working version you have at hand)?

**Comment by cproc** [comment]
The hardware is an Intel NUC 12 (Alder Lake-P). According to `/report/drivers/devices` the IO memory range belongs to the Intel GPU (@jschlatow: do you still need more information from the report?).

**Comment by jschlatow** [comment]
Thanks. That's what I suspected.

What's puzzling me is that intel_gpu_drv and boot_fb are both started alongside. The sculpt manager actually considers them mutually exclusive.

**Comment by nfeske** [comment]
> the download of packages into the ram_fs, when used as depot source, reliable stucks with large packets

I guess that commit https://github.com/genodelabs/genode/commit/c35c71a6615578efe6ce19bdf511908fa9192441 is the troublemaker here. I should definitely add a warning once the limit is reached and relax the limit for ram_fs. @alex-ab do you have an intuition what a sensible limit could be?

**Comment by cproc** [comment]
> What's puzzling me is that intel_gpu_drv and boot_fb are both started alongside. The sculpt manager actually considers them mutually exclusive.

I instrumented `Sculpt::Fb_driver::update()` now and saw 3 calls in the log. On the first call `use_intel` was false and on the later two calls `use_intel` was true.

**Comment by nfeske** [comment]
@alex-ab may you give e424fc59d4947ed4b34527330d72acde759df44f a try?

**Comment by alex-ab** [comment]
> @alex-ab may you give [e424fc5](https://github.com/genodelabs/genode/commit/e424fc59d4947ed4b34527330d72acde759df44f) a try?

Works, splendid :+1:

**Comment by alex-ab** [comment]
> What's puzzling me is that intel_gpu_drv and boot_fb are both started alongside. The sculpt manager actually considers them mutually exclusive.

Even so it seems a configuration issue, would it be sensible to catch such attempts in the platform driver, so that it survives. It may not fulfill the service for the specific driver, but stays alive for all other in the system ?

**Comment by jschlatow** [comment]
> Even so it seems a configuration issue, would it be sensible to catch such attempts in the platform driver, so that it survives. It may not fulfill the service for the specific driver, but stays alive for all other in the system ?

That's a reasonable suggestion. I'll have a look.

Regarding the sculpt-manager issue: It seems there is some sort of a race. The sculpt manager seems to get to work before the devices ROM is complete, i.e. on the first call of `Sculpt::Fb_driver::update()` the Intel display controller is not present and therefore boot_fb is started. It seems though that the Sculpt manager does not (yet) implement a handover from boot_fb to intel_fb since the start node for boot_fb is never removed once it has been constructed. @nfeske What is the rationale for not using `_boot_fb.conditional(use_boot_fb, ...)` in _sculpt_manager/driver/fb.h_?

**Comment by jschlatow** [comment]
> Could we consider the platform driver to defer the reporting the `boot_fb` device until the point where PCI info is also reported? This would avoid the need to swapping out drivers, which is a new behavior that is possibly best to avoid for the time being.

If I'm not mistaken, the boot_fb is not part of the devices ROM but is announced inthe platform_info ROM. @cproc Could you have a look what the content of the devices ROM is when `Sculpt::Board_info::Detected::from_xml()` is called the first time? I'm wondering whether the devices ROM is empty (which would be a good indicator to skip further evaluation and wait for the next update).

**Comment by cproc** [comment]
> @cproc Could you have a look what the content of the devices ROM is when `Sculpt::Board_info::Detected::from_xml()` is called the first time? I'm wondering whether the devices ROM is empty (which would be a good indicator to skip further evaluation and wait for the next update).

Yes, the devices ROM is empty on the first call. The driver manager had a check for that:

https://github.com/genodelabs/genode/blob/9d19410f12e3e4b851e46d56c30749bf84598c6e/repos/gems/src/app/driver_manager/main.cc#L457-L463

**Comment by jschlatow** [comment]
> > Even so it seems a configuration issue, would it be sensible to catch such attempts in the platform driver, so that it survives. It may not fulfill the service for the specific driver, but stays alive for all other in the system ?
> 
> That's a reasonable suggestion. I'll have a look.

@cproc Can you give 6bba9a4 a try and see whether the platform driver survives the failure to create an IO_MEM session?

**Comment by nfeske** [comment]
Thanks @cproc for investigating. It is probably best to add the check to the sculpt manager then (along with a comment on why it is needed).

**Comment by alex-ab** [comment]
It was, up to now, not supported to use boot_fb with intel_gpu -  either you use intel gpu + display driver _or_ boot_fb. Of course, this must not stay that way, is just untested/never used.

**Comment by cproc** [comment]
> @cproc Can you give [6bba9a4](https://github.com/genodelabs/genode/commit/6bba9a47742c7592943197d2f028066a5d47f597) a try and see whether the platform driver survives the failure to create an IO_MEM session?

Yes, it survives.

**Comment by nfeske** [comment]
@cnuke just pointed out to me that the prepare step upon the selection of a used partition is not always executed. Just noting that I am able to reproduce it. I'm on it.

**Comment by nfeske** [comment]
> used partition is not always executed [...]

Fixed by https://github.com/genodelabs/genode/commit/75b5b8a37ee1a2cb4088022cb0c87baa73c15d5f

**Comment by nfeske** [comment]
> Do ARM platforms get a valid devices ROM too?

They should.

However, since this test is for the specific reason of reliably detecting boot_fb vs intel, wouldn't it be sensible to test the condition in fb.h? For the other devices (storage, usb) such ambiguities do not exist.

**Comment by cnuke** [comment]
> > used partition is not always executed [...]
> 
> Fixed by [75b5b8a](https://github.com/genodelabs/genode/commit/75b5b8a37ee1a2cb4088022cb0c87baa73c15d5f)

Thanks for the fix, works nicely :+1:.

**Comment by cproc** [comment]
> However, since this test is for the specific reason of reliably detecting boot_fb vs intel, wouldn't it be sensible to test the condition in fb.h? For the other devices (storage, usb) such ambiguities do not exist.

`fb.h` currently only knows `Board_info` and the platform ROM. Without changing the interface, the check could be added to `Sculpt::Board_info::Detected::from_xml()` and skip the `boot_fb` detection if the devices ROM is empty or it could be added to `Sculpt::Drivers::Instance::_handle_devices(Xml_node const &devices)` and skip the `_fb_driver` update if the devices ROM is empty.

**Comment by nfeske** [comment]
@cproc sorry for stirring up confusion. Your very last suggestion sounds best to me.

**Comment by cproc** [comment]
Commit 8410bc0 adds the devices ROM check.

**Comment by nfeske** [comment]
@cproc Perfect! Merged to staging.

**Comment by mewmew** [comment]
Perhaps this is working as intended. But I noticed a difference in how the config file system is handled between Sculpt 23.10 and 24.04.

In particular, using 23.10, it was possible to create a persistent config directory on e.g. the used USB drive, such that those configuration files would be loaded when the Genode partition of the USB drive was used.

Personally, I've used this to add new launchers to Sculpt without having to edit e.g. the `default-pc.sculpt` file. Those launchers would be placed in `/usb-1.3.3/config/23.10/launchers/xxx`.

I tried to use the same setup for Sculpt 24.04 (using `/usb-1.3.3/config/24.04/launchers/xxx` at revision b9bd93847baa1d25416138c8dd93b09440583a18), but it doesn't seem to pick up the changes. I expected the new launchers to be made available through the new `+ -> Options` menu, but they don't seem to be there. If I manually copy the files from `/usb-1.3.3/config/24.04/launchers/xxx` to `/config/launcher/xxx` then the new launchers appear under `+ -> Options` and work correctly.

Is this the intended behaviour? Can I still use persistent config files (and launchers) on Sculpt 24.04? Perhaps in some other way?

Cheers,
Robin

**Comment by nfeske** [comment]
> But I noticed a difference in how the config file system is handled between Sculpt 23.10 and 24.04.

This is not intended and is most likely related to the symptom mentioned at comment https://github.com/genodelabs/genode/issues/5174#issuecomment-2049826341. The original behavior should hopefully be restored by commit https://github.com/genodelabs/genode/commit/75b5b8a37ee1a2cb4088022cb0c87baa73c15d5f (on staging).

**Comment by nfeske** [comment]
@mewmew it now occurred to me that you were already on the latest staging yesterday. Let me investigate...

**Comment by nfeske** [comment]
@mewmew unfortunately I'm not able to reproduce the persistent-config problem. I tested the mechanism with a partition on an NVMe device and also a USB stick. My custom launcher and custom nic_router config are imported from the respective config/24.04/ directory just fine. Do you see the "prepare" component in the graph after selecting your Genode partition to "Use"? (you may also look at /report/log to see if the prepare step was executed)

**Comment by mewmew** [comment]
> @mewmew unfortunately I'm not able to reproduce the persistent-config problem. I tested the mechanism with a partition on an NVMe device and also a USB stick. My custom launcher and custom nic_router config are imported from the respective config/24.04/ directory just fine. Do you see the "prepare" component in the graph after selecting your Genode partition to "Use"? (you may also look at /report/log to see if the prepare step was executed)

It seems to work now! A few errors appear in the log related to the `prepare` component, but they can probably be ignored as it does succeed to populate the `/config/launcher` files.

Sorry for the confusion, I am guessing the behaviour I observed earlier was prior to your 75b5b8a37ee1a2cb4088022cb0c87baa73c15d5f commit. I must have mixed up the revisions when running the test.

Thanks for the quick solution!

(Just for reference, the errors in the log related to the `prepare` component are as follows:)
```
runtime -> prepare -> bash   Error: no plugin foud for fcntl(3)
...
runtime -> prepare -> bash   Error: no plugin foud for fcntl(19)
```

**Comment by nfeske** [comment]
@mewmew the bash-related error messages in the log are normal. They admittedly spoil the aesthetics of the log but are not critical. As I'm planning to simplify the prepare step (using fs_tool instead of bash) down the road anyway, those ugly lines will vanish sometime later.

**Comment by cnuke** [comment]
Commit [5ef8105](https://github.com/cnuke/genode/commit/5ef810519171ad1d2091a9217d88e0112965f143) and [ceb1286](https://github.com/cnuke/genode/commit/ceb1286609961dc76833f7ab842b05423c090ee9) allow for driving a UHD display.

However, to interact with the leitzentrale - the `runtime_view` in particular - commit [0e6c126](https://github.com/cnuke/genode/commit/0e6c126322e45ce4d6fc7d8492e862f5f84322cd) is necessary. Although the `runtime_view` gets upgrade eventually it takes multiple attempts and between the various upgrade attempts the whole leitzentrale is sometimes frozen and gets restarted. The quota of `52M` is the value that is required to use font-size large (w/o any upgrade).

**Comment by nfeske** [comment]
Thank you @cnuke for the high-resolution adjustments. I agree with the changes and find the increase of the runtime_view quota sensible. It hosts all dialogs after all.

To counter the total increase of the leitzentrale quotas a bit, we could consider reducing the quota assigned to the drivers subsystem as a subsequent step. With the move of most driver components to the runtime, it got a much smaller footprint.

**Comment by nfeske** [comment]
@cproc while testing your latest build of falkon-jemalloc, I noticed that the log gets spammed with messages like this:
```
falkon-jemalloc -> falkon] Error: fcntl(): command 0xc not supported - vfs
```
Can you reproduces this? Could those messages be silenced?

**Comment by nfeske** [comment]
@cnuke, the mixer launcher should be improved a little:

* I noticed that the both play policies - the one for vbox and the default rule - match my vbox instance. So the vbox output is incorporated into the mix two times. I would comment out the vbox rules, turn their labels to fully qualified `label` (w/o suffix) and add a comment that the fully qualified rule is an alternative to the label-suffix wildcard. So the uninitiated user gets a working mixer but an advanced user can be more specific by replacing the wildcard.

* The default mixer config should leave the microphone disconnected. The rule of enabling the mix should be there in form of a comment. This way, we can safely add the mixer to the falkon-preset, connect both play and record sessions of the browser to the mixer, and meet the user's expectation that no mic data is exposed to the browser unless explicitly enabled.

**Comment by nfeske** [comment]
I've update the falkon preset for the use of the mixer in https://github.com/genodelabs/genode/commit/c422c8cfddba225eb47c31f23269bf5531ae9726. The audio driver is not started by default. So one has to enable the 'audio' option.

@cproc @cnuke the falkon pkg by cproc lacks the oss attributes that improve the audio stability. When using the falkon browser for watching a youtube video, there are currently quite irritating garbling sounds. @cproc could you give the preset a spin and possible update the pkg with the adjusted oss settings?

**Comment by nfeske** [comment]
@chelmuth reported the following keyboard focus issue: Unless an inspect view is open, on cannot edit text via the text editor in the files tab. I'm on this one.

**Comment by nfeske** [comment]
Commit https://github.com/genodelabs/genode/commit/9d2e389358472de1084eebab707dbba30773981d should fix the keyboard-focus issue.

**Comment by cproc** [comment]
> @cproc while testing your latest build of falkon-jemalloc, I noticed that the log gets spammed with messages like this:
> 
> ```
> falkon-jemalloc -> falkon] Error: fcntl(): command 0xc not supported - vfs
> ```
> 
> Can you reproduces this? Could those messages be silenced?

Command 0xc ist `F_SETLK`. There are multiple places in the Qt source code where this command appears. Should we add the command to the libc VFS plugin as dummy which always succeeds to silence the messages?

**Comment by nfeske** [comment]
@m-stein commit https://github.com/genodelabs/genode/commit/f0709f3f53c38d1b82f9d6255c4dcb409aebb89f fixes the keyboard-layout selection issue you reported offline.

**Comment by nfeske** [comment]
> Should we add the command to the libc VFS plugin as dummy which always succeeds to silence the messages?

Or could we print the error just once and suppress subsequent messages referring to the same argument as the one last printed? So the log would reveal the values not handled but large sequences of the same diagnostic messages get somewhat compressed into one.

**Comment by chelmuth** [comment]
> Commit [9d2e389](https://github.com/genodelabs/genode/commit/9d2e389358472de1084eebab707dbba30773981d) should fix the keyboard-focus issue.

Not reproducible anynmore, thanks.

**Comment by cnuke** [comment]
> @cnuke, the mixer launcher should be improved a little:
> 
> […]

Commit [ab5d44c](https://github.com/cnuke/genode/commit/ab5d44c2ffd052847ed92b37aba0696585ba3d29) addresses your remarks while commit [c6c7221](https://github.com/cnuke/genode/commit/c6c7221429a35956492c2a2f90fe5ff4eb6619f1) and [4b73e28](https://github.com/cnuke/genode/commit/4b73e28807197cc35bab6ce8162c1ea1db90260b) refine the state-report handling in the mixer (I am not overly happy with the `Operations` change but it is most straight-forward).

**Comment by chelmuth** [comment]
Commit b58ddd2 probes the i8042 on PC ps_drv startup and enters infinite sleep if no controller responded.

**Comment by nfeske** [comment]
Thanks @cproc for the fcntl patch and the updated falkon pkgs. I merged the libc patch to staging. Fixup b198059ce480545842e959c4d2b1dbf960b83374 updates the browser version of the falkon preset.

**Comment by nfeske** [comment]
@cnuke thanks a lot for the mixer adjustments. The changes look all good to me.

**Comment by nfeske** [comment]
While testing the falkon preset, the browser produces messages like the following:

```
Error: could not open ROM session for "libresolv.so"
...
Error: could not open ROM session for "resolv"
```

The browser continues working but something might be wrong.

When browsing https://github.com/genodelabs/genode in falkon, the directories are not shown. May this effect be related to the error messages?

**Comment by cproc** [comment]
> When browsing https://github.com/genodelabs/genode in falkon, the directories are not shown. May this effect be related to the error messages?

I can reproduce the GitHub directories problem on Linux too. It does not occur with newer Qt versions (with newer Chromium engine). I also found a bug report on GitHub which seems to describe this problem: https://github.com/orgs/community/discussions/87019

**Comment by nfeske** [comment]
Commit https://github.com/genodelabs/genode/commit/ea6f2a92478f5841d4b98245378a347a387937e8 restores the automatic restart of intel_fb whenever its heart fails beating.

**Comment by jschlatow** [comment]
@nfeske It's probably not a new one but since we dedicated this year to usability: The file view behaves oddly when scrolling with the mouse wheel. If I scroll through _/report/log_ with page down, I end up with the last 39 lines or so. If I use the mouse wheel, however, I can scroll further down. This has the effect that the number of lines shown is decreased up until there is no line and only a very small and empty square.

**Comment by nfeske** [comment]
> If I use the mouse wheel, however, I can scroll further down.

@jschlatow thanks for reporting. Commit https://github.com/genodelabs/genode/commit/e505e9e941d0fa0f850560b72ac516b0392a25ad fixes this regression.

**Comment by cnuke** [comment]
@nfeske commit [0776f4b](https://github.com/cnuke/genode/commit/0776f4b738b1ed1e112a23d4ffa2ff67f7eb63f3) adds the interim fix for SoC fb selection.

**Comment by nfeske** [comment]
> commit [0776f4b](https://github.com/cnuke/genode/commit/0776f4b738b1ed1e112a23d4ffa2ff67f7eb63f3) adds the interim fix for SoC fb selection.

@cnuke thank you for the fix and for commenting the condition so well.

**Comment by cnuke** [comment]
@nfeske commit [ca7fed4](https://github.com/cnuke/genode-allwinner/commit/ca7fed434eaf3b7a7cc1b14a44e4bf1d4c827171) allows for switching to the runtime again. It is marked as *WIP* as I merely switched the sessions and have not checked for any repercussion from doing so (but allowed me to test other things).

**Comment by nfeske** [comment]
> It is marked as WIP as I merely switched the sessions

The change is definitely valid. Hence, I merged your commit (with the WIP removed) right away. I very much appreciate it!

**Comment by skalk** [comment]
@nfeske commits 0bf821a and 7c7ea03 are necessary to run Sculpt OS on the MNT reform2 after the recent changes.

**Comment by jschlatow** [comment]
@nfeske I've been trying to apply my Sculpt customisations to 24.04. I noticed that I lose my mouse and/or keyboard input when I set any `event_filter` in my Sculpt config. With `event_filter: pc` I get neither mouse nor keyboard input. With `event_filter: default` I only get mouse input. Only if I omit the `event_filter` line, I get mouse and keyboard input.

**Comment by nfeske** [comment]
@jschlatow could you compare these two event-filter configs with the one generated by the Sculpt manager under config/managed/event_filter?

**Comment by jschlatow** [comment]
> @jschlatow could you compare these two event-filter configs with the one generated by the Sculpt manager under config/managed/event_filter?

Got it (fixed by 3e80e70). I've compared the files by eye before but only noticed the missing "keyboard/" before the included ROM names. vimdiff-ing the file actually revealed that the ".chargen" needed to be removed as well. I have no clue, however, why using the wrong ROM names still worked until 23.10.

**Comment by jschlatow** [comment]
@nfeske I noticed another thing: When the list of launchers becomes quite long, the Add/Options buttons in the "+" menu are placed behind the Files/Components/Inspect menu.

**Comment by jschlatow** [comment]
@nfeske: I'm afraid I found another bug: I just tested the goa_testbed preset. On my customised Sculpt, I have added a "system_info" launcher. This launcher specifies routes to a "report_fs_rom" child. The goa_testbed preset, however, routes the ROM requests to the "fs_rom" child. When I switch to the goa_testbed preset, the Leitzentrale complains with `system_info requires report_fs_rom`, which is not correct because the deploy config correctly specifies the route to "fs_rom".

**Comment by jschlatow** [comment]
@chelmuth I updated the goa_testbed preset in 5a5d5ed.

**Comment by nfeske** [comment]
> The goa_testbed preset, however, routes the ROM requests to the "fs_rom" child.

Have you reviewed the routing rules in config/managed/runtime?

**Comment by nfeske** [comment]
> The system_info is missing in /config/managed/runtime.

You say that system_info is featured in managed/deploy but not in managed/runtime? Then it looks like some condition is not satisfied (e.g., pkg not installed, or a missing ROM). If so, a look into report/runtime/depot_query/blueprint might give us further insights. Are there information about the system_info pkg present?

**Comment by nfeske** [comment]
Does managed/deploy happens to have two start nodes for system_info, one referring to the launcher and one referring the a pkg?

**Comment by jschlatow** [comment]
> Does managed/deploy happens to have two start nodes for system_info, one referring to the launcher and one referring the a pkg?

No, there is only a single start node for system_info.

**Comment by jschlatow** [comment]
> > Does managed/deploy happens to have two start nodes for system_info, one referring to the launcher and one referring the a pkg?
> 
> No, there is only a single start node for system_info.

I tried a few other things and noticed that the situation is triggered, when I had enabled the system_info launcher before and then switch to the preset. I believe the Sculpt manager tries to keep the already existing child instead of replacing it with the once from the preset.

**Comment by nfeske** [comment]
> I believe the Sculpt manager tries to keep the already existing child instead of replacing it with the one from the preset.

That is perfectly plausible! Whenever name and version remain the same, the sculpt manager retains the start node. As a quick and dirty fix, you may try to set the version attribute in the start node of the preset to some distinct value. To fix this situation in profound way, I need to dig deeper.

**Comment by jschlatow** [comment]
> > I believe the Sculpt manager tries to keep the already existing child instead of replacing it with the one from the preset.
> 
> That is perfectly plausible! Whenever name and version remain the same, the sculpt manager retains the start node. As a quick and dirty fix, you may try to set the version attribute in the start node of the preset to some distinct value. To fix this situation in profound way, I need to dig deeper.

Adding a version attribute in the preset does not make a difference.

**Comment by nfeske** [comment]
> I noticed another thing: When the list of launchers becomes quite long, the Add/Options buttons in the "+" menu are placed behind the Files/Components/Inspect menu.

I'm able to reproduce it but isn't it fine as is? You can reach the Add/Option tabs by scrolling the popup using the scroll wheel or via the page-up/down keys. As the popup belongs to the context of the "Components" tab, it should in my opinion not overlap the panel. The panel should be stronger so to say.

**Comment by nfeske** [comment]
@jschlatow sorry, I meant commit 6cac8293ed13bea00e04ae337cce252a4a4f4d29.

**Comment by cnuke** [comment]
Commit [9f83d47](https://github.com/cnuke/genode/commit/9f83d472cea4776e2b644341bd07d7ac7fdd34b9) adjust the default driver's volume to the max.

**Comment by nfeske** [comment]
Commit https://github.com/genodelabs/genode/commit/4c0cd599aaa8c498da68444e547b1096e558a4ad renames the event-filter input "touch" to "touchpad" to disambiguate the term against touch screens.

**Comment by jschlatow** [comment]
> @jschlatow sorry, I meant commit [6cac829](https://github.com/genodelabs/genode/commit/6cac8293ed13bea00e04ae337cce252a4a4f4d29).

Works like a charm.

**Comment by jschlatow** [comment]
> > Adding a version attribute in the preset does not make a difference.
> 
> Then the only other short-term work-around I see would be naming either your launcher or the system_info start node different to avoid the ambiguity. I suggest this band-aid because I would like to defer the work on a real fix to the time after the release as I currently need to focus on the documentation. The (wrong) behavior is not a regression after all.

Another interim solution is to first switch to the _empty_ preset.

**Comment by jschlatow** [comment]
@nfeske I found another, critical, bug in the Sculpt UI. I tried installing the file vault from Martin's index. In the dialog where one picks the routes, I clicked the yellow triangle because I wanted to go back to the previous dialog. This interaction reproducibly triggers a page fault and freezes the entire system.

**Comment by nfeske** [comment]
@jschlatow as I'm currently intensively working on the documentation, would you be willing to investigate this?

**Comment by jschlatow** [comment]
> @jschlatow as I'm currently intensively working on the documentation, would you be willing to investigate this?

I will have a look. Let me know if you have any pointers.

**Comment by jschlatow** [comment]
> > I will have a look. Let me know if you have any pointers.
> 
> I'm currently co0-investigating the issue while @nfeske still updates the documentation. It seems the issue correlates with the fact that mstein/index does not use `<index>` elements.

That makes sense, the instruction that triggers the page fault seems to be somewhere in the XML parsing code.

**Comment by chelmuth** [comment]
Commit 61ac7f8229124c561a1cd6dae2282718f8f4008d fixes the popup dialog crash.

**Comment by jschlatow** [comment]
Can anyone reproduce an audio echo, i.e. the recorded microphone input being played? Starting the default mixer and audio launchers on the x260 or t490s reproduces this echo. It's most obvious when I touch or stroke the laptop lid.

**Comment by chelmuth** [comment]
@ssumpf I integrated preliminary touchpad support into staging. The actual driver is left out and taken as binary pkg from my depot. Please give it a try on U7511, but don't miss to adapt _/config/launcher/touchpad_ for your device.

**Comment by nfeske** [comment]
> Can anyone reproduce an audio echo

With the current version of the mixer launcher, this should not longer be the case. The driver's play sessions for the mic should be labeled as mic_left and mic_right. Can you take a look at report/runtime/mixer/state to see if this is the case?

I can deliberately reproduce an echo on my x250 via the mixer by adding corresponding `<play>` nodes to the wildcard rules for left and right (stroking the lid produces a nice effect). But it is not by default.

May the audio driver on your machines bridge the mic input to the output (as some kind of monitor mode by the hardware)?

**Comment by nfeske** [comment]
> I integrated preliminary touchpad support into staging. 

This works splendidly on my Gen12 Framework. Thanks a lot @chelmuth!

**Comment by jschlatow** [comment]
> > If I use the mouse wheel, however, I can scroll further down.
> 
> @jschlatow thanks for reporting. Commit [e505e9e](https://github.com/genodelabs/genode/commit/e505e9e941d0fa0f850560b72ac516b0392a25ad) fixes this regression.

@nfeske I just noticed that the issue is still present when viewing the _log_ file.

**Comment by mewmew** [comment]
> I will have a look. Let me know if you have any pointers.

`0x3a28213a`
`0x6339392c`
`0x7363682e`

[I couldn't help it.](https://xkcd.com/138/)

**Comment by ssumpf** [comment]
> @ssumpf I integrated preliminary touchpad support into staging. The actual driver is left out and taken as binary pkg from my depot. Please give it a try on U7511, but don't miss to adapt _/config/launcher/touchpad_ for your device.

Up an running :+1:

**Comment by jschlatow** [comment]
> With the current version of the mixer launcher, this should not longer be the case. The driver's play sessions for the mic should be labeled as mic_left and mic_right. Can you take a look at report/runtime/mixer/state to see if this is the case?

I'm running the most recent config (except the `warning_rate_ms` change). However, the mixer's state indicates that the record and play sessions have the same labels ("audio -> right" and "audio -> left"). Strangely, this is the case for both devices (x260 and t490s).

**Comment by jschlatow** [comment]
@nfeske I just double-checked both issues (scrolling in file view and audio echo) with the rc1 image and cannot reproduce them. Will try rebuilding my archives.

**Comment by jschlatow** [comment]
> @nfeske I just double-checked both issues (scrolling in file view and audio echo) with the rc1 image and cannot reproduce them. Will try rebuilding my archives.

I confirm that rebuilding my depot fixed both issues. Sorry for the noise.

**Comment by nfeske** [comment]
Commit https://github.com/genodelabs/genode/commit/69cef8428bbfb4cc49dea926f5264c406d2dc942 updates the documentation. Many thanks to @jschlatow and @cproc for the super careful review.

**Comment by chelmuth** [comment]
@atopia I merged your inspect adjustment as fixup d55f3a4a548338375ec2d7c9f4ce2b2a9bd7bbf7.

**Comment by nfeske** [comment]
Thanks to everyone involved! It's wonderful how the release came together.
